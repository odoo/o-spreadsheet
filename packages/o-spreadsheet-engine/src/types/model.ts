import { GeoChartRegion } from "./chart/geo_chart";
import { Client, ClientPosition } from "./collaborative/session";
import { TransportService } from "./collaborative/transport_service";
import { Currency } from "./currency";
import { InformationNotification } from "./env";
import { FileStore } from "./files";
import { Locale } from "./locale";
import { Color } from "./misc";

/**
 * Model
 *
 * The Model class is the owner of the state of the Spreadsheet. However, it
 * has more a coordination role: it defers the actual state manipulation work to
 * plugins.
 *
 * At creation, the Model instantiates all necessary plugins. They each have
 * a private state (for example, the Selection plugin has the current selection).
 *
 * State changes are then performed through commands.  Commands are dispatched
 * to the model, which will then relay them to each plugins (and the history
 * handler). Then, the model will trigger an 'update' event to notify whoever
 * is concerned that the command was applied (if it was not cancelled).
 *
 * Also, the model has an unconventional responsibility: it actually renders the
 * visible viewport on a canvas. This is because each plugins actually manage a
 * specific concern about the content of the spreadsheet, and it is more natural
 * if they are able to read data from their internal state to represent it on the
 * screen.
 *
 * Note that the Model can be used in a standalone way to manipulate
 * programmatically a spreadsheet.
 */
export type Mode = "normal" | "readonly" | "dashboard";

export interface ModelConfig {
  readonly mode: Mode;
  /**
   * Any external custom dependencies your custom plugins or functions might need.
   * They are available in plugins config and functions
   * evaluation context.
   */
  readonly custom: Readonly<{
    [key: string]: any;
  }>;
  readonly defaultCurrency?: Partial<Currency>;
  /**
   * External dependencies required to enable some features
   * such as uploading images.
   */
  readonly external: Readonly<ModelExternalConfig>;
  readonly moveClient: (position: ClientPosition) => void;
  readonly transportService: TransportService;
  readonly client: Client;
  readonly snapshotRequested: boolean;
  readonly notifyUI: (payload: InformationNotification) => void;
  readonly raiseBlockingErrorUI: (text: string) => void;
  readonly customColors: Color[];
}

export interface ModelExternalConfig {
  readonly fileStore?: FileStore;
  readonly loadCurrencies?: () => Promise<Currency[]>;
  readonly loadLocales?: () => Promise<Locale[]>;
  readonly geoJsonService?: {
    getAvailableRegions: () => GeoChartRegion[];
    getTopoJson: (region: string) => Promise<any>;
    /**  Convert the name of a geographical feature (eg. France) to the id of the corresponding feature in the TopoJSON */
    geoFeatureNameToId: (region: string, territory: string) => string | undefined;
  };
}
