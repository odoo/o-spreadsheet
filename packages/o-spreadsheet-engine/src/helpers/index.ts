export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
