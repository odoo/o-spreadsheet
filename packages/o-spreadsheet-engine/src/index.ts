export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
