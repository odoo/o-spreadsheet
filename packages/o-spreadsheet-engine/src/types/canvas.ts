export type CanvasSurface = HTMLCanvasElement | OffscreenCanvas;
export type Canvas2DContext = CanvasRenderingContext2D | OffscreenCanvasRenderingContext2D;
