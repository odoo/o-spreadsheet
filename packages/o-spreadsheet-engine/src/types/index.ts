export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
