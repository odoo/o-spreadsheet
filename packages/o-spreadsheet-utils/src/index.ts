export * from "./translation";
export * from "./uuid";
