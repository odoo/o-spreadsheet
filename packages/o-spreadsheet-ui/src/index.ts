import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}