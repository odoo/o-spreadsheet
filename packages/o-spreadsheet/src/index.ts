import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };