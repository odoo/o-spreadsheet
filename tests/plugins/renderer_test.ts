import { Model } from "../../src/model";
import { MockCanvasRenderingContext2D } from "../canvas.mock";
import { Viewport, GridRenderingContext } from "../../src/types";
import { toZone } from "../../src/helpers";

MockCanvasRenderingContext2D.prototype.measureText = function () {
  return { width: 100 };
};

interface ContextObserver {
  onSet?(key, val): void;
  onGet?(key): void;
  onFunctionCall?(fn: string, args: any[]): void;
}

class MockGridRenderingContext implements GridRenderingContext {
  _context = document.createElement("canvas").getContext("2d");
  ctx: CanvasRenderingContext2D;
  viewport: Viewport;
  dpr = 1;
  thinLineWidth = 0.4;

  constructor(model: Model, width: number, height: number, observer: ContextObserver) {
    this.viewport = model.getters.getAdjustedViewport(
      {
        width,
        height,
        offsetX: 0,
        offsetY: 0,
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
      },
      "zone"
    );

    const handler = {
      get: (target, val) => {
        if (val in (this._context as any).__proto__) {
          return (...args) => {
            if (observer.onFunctionCall) {
              observer.onFunctionCall(val, args);
            }
          };
        } else {
          if (observer.onGet) {
            observer.onGet(val);
          }
        }
        return target[val];
      },
      set: (target, key, val) => {
        if (observer.onSet) {
          observer.onSet(key, val);
        }
        target[key] = val;
        return true;
      },
    };
    this.ctx = new Proxy({}, handler);
  }
}

describe("renderer", () => {
  test("snapshot for a simple grid rendering", () => {
    const model = new Model();

    model.dispatch("SET_VALUE", { xc: "A1", text: "1" });
    const instructions: string[] = [];
    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        instructions.push(`context.${key}=${JSON.stringify(value)};`);
      },
      onGet: (key) => {
        instructions.push(`GET:${key}`);
      },
      onFunctionCall: (key, args) => {
        instructions.push(`context.${key}(${args.map((a) => JSON.stringify(a)).join(", ")})`);
      },
    });

    model.drawGrid(ctx);
    expect(instructions).toMatchSnapshot();
  });

  test("formulas evaluating to a string are properly aligned", () => {
    const model = new Model();

    model.dispatch("SET_VALUE", { xc: "A1", text: "1" });
    model.dispatch("SET_VALUE", { xc: "A2", text: "=A1" });

    let textAligns: string[] = [];
    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        if (key === "textAlign") {
          textAligns.push(value);
        }
      },
    });

    model.drawGrid(ctx);
    expect(textAligns).toEqual(["right", "right", "center"]); // center for headers

    textAligns = [];
    model.dispatch("SET_VALUE", { xc: "A1", text: "asdf" });
    model.drawGrid(ctx);
    expect(textAligns).toEqual(["left", "left", "center"]); // center for headers
  });

  test("formulas in a merge, evaluating to a string are properly aligned", () => {
    const model = new Model();

    model.dispatch("ADD_MERGE", { sheet: "Sheet1", zone: toZone("A2:B2") });
    model.dispatch("SET_VALUE", { xc: "A1", text: "1" });
    model.dispatch("SET_VALUE", { xc: "A2", text: "=A1" });

    let textAligns: string[] = [];

    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        if (key === "textAlign") {
          textAligns.push(value);
        }
      },
    });
    model.drawGrid(ctx);

    expect(textAligns).toEqual(["right", "right", "center"]); // center for headers

    model.dispatch("SET_VALUE", { xc: "A1", text: "asdf" });

    textAligns = [];
    model.drawGrid(ctx);
    expect(textAligns).toEqual(["left", "left", "center"]); // center for headers
  });

  test("formulas evaluating to a boolean are properly aligned", () => {
    const model = new Model();

    model.dispatch("SET_VALUE", { xc: "A1", text: "1" });
    model.dispatch("SET_VALUE", { xc: "A2", text: "=A1" });

    let textAligns: string[] = [];
    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        if (key === "textAlign") {
          textAligns.push(value);
        }
      },
    });
    model.drawGrid(ctx);
    expect(textAligns).toEqual(["right", "right", "center"]); // center for headers

    textAligns = [];
    model.dispatch("SET_VALUE", { xc: "A1", text: "true" });
    model.drawGrid(ctx);
    expect(textAligns).toEqual(["center", "center", "center"]); // center for headers
  });

  test("formulas in a merge, evaluating to a boolean are properly aligned", () => {
    const model = new Model();

    model.dispatch("ADD_MERGE", { sheet: "Sheet1", zone: toZone("A2:B2") });
    model.dispatch("SET_VALUE", { xc: "A1", text: "1" });
    model.dispatch("SET_VALUE", { xc: "A2", text: "=A1" });

    let textAligns: string[] = [];

    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        if (key === "textAlign") {
          textAligns.push(value);
        }
      },
    });
    model.drawGrid(ctx);

    expect(textAligns).toEqual(["right", "right", "center"]); // center for headers

    model.dispatch("SET_VALUE", { xc: "A1", text: "false" });

    textAligns = [];
    model.drawGrid(ctx);
    expect(textAligns).toEqual(["center", "center", "center"]); // center for headers
  });

  test("errors are aligned to the center", () => {
    const model = new Model();

    model.dispatch("SET_VALUE", { xc: "A1", text: "=A1" });

    let textAligns: string[] = [];

    let ctx = new MockGridRenderingContext(model, 1000, 1000, {
      onSet: (key, value) => {
        if (key === "textAlign") {
          textAligns.push(value);
        }
      },
    });
    model.drawGrid(ctx);

    // 1 center for headers, 1 for cell content
    expect(textAligns).toEqual(["center", "center"]);
  });
});
