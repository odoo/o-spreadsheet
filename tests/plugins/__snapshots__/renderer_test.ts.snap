// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer snapshot for a simple grid rendering 1`] = `
Array [
  "context.fillStyle=\\"white\\";",
  "context.fillRect(0, 0, 1000, 1000)",
  "context.lineWidth=0.16000000000000003;",
  "context.strokeStyle=\\"#222\\";",
  "context.beginPath()",
  "context.moveTo(156, 0)",
  "context.lineTo(156, 1000)",
  "context.moveTo(252, 0)",
  "context.lineTo(252, 1000)",
  "context.moveTo(348, 0)",
  "context.lineTo(348, 1000)",
  "context.moveTo(444, 0)",
  "context.lineTo(444, 1000)",
  "context.moveTo(540, 0)",
  "context.lineTo(540, 1000)",
  "context.moveTo(636, 0)",
  "context.lineTo(636, 1000)",
  "context.moveTo(732, 0)",
  "context.lineTo(732, 1000)",
  "context.moveTo(828, 0)",
  "context.lineTo(828, 1000)",
  "context.moveTo(924, 0)",
  "context.lineTo(924, 1000)",
  "context.moveTo(1020, 0)",
  "context.lineTo(1020, 1000)",
  "context.moveTo(0, 49)",
  "context.lineTo(1000, 49)",
  "context.moveTo(0, 72)",
  "context.lineTo(1000, 72)",
  "context.moveTo(0, 95)",
  "context.lineTo(1000, 95)",
  "context.moveTo(0, 118)",
  "context.lineTo(1000, 118)",
  "context.moveTo(0, 141)",
  "context.lineTo(1000, 141)",
  "context.moveTo(0, 164)",
  "context.lineTo(1000, 164)",
  "context.moveTo(0, 187)",
  "context.lineTo(1000, 187)",
  "context.moveTo(0, 210)",
  "context.lineTo(1000, 210)",
  "context.moveTo(0, 233)",
  "context.lineTo(1000, 233)",
  "context.moveTo(0, 256)",
  "context.lineTo(1000, 256)",
  "context.moveTo(0, 279)",
  "context.lineTo(1000, 279)",
  "context.moveTo(0, 302)",
  "context.lineTo(1000, 302)",
  "context.moveTo(0, 325)",
  "context.lineTo(1000, 325)",
  "context.moveTo(0, 348)",
  "context.lineTo(1000, 348)",
  "context.moveTo(0, 371)",
  "context.lineTo(1000, 371)",
  "context.moveTo(0, 394)",
  "context.lineTo(1000, 394)",
  "context.moveTo(0, 417)",
  "context.lineTo(1000, 417)",
  "context.moveTo(0, 440)",
  "context.lineTo(1000, 440)",
  "context.moveTo(0, 463)",
  "context.lineTo(1000, 463)",
  "context.moveTo(0, 486)",
  "context.lineTo(1000, 486)",
  "context.moveTo(0, 509)",
  "context.lineTo(1000, 509)",
  "context.moveTo(0, 532)",
  "context.lineTo(1000, 532)",
  "context.moveTo(0, 555)",
  "context.lineTo(1000, 555)",
  "context.moveTo(0, 578)",
  "context.lineTo(1000, 578)",
  "context.moveTo(0, 601)",
  "context.lineTo(1000, 601)",
  "context.moveTo(0, 624)",
  "context.lineTo(1000, 624)",
  "context.moveTo(0, 647)",
  "context.lineTo(1000, 647)",
  "context.moveTo(0, 670)",
  "context.lineTo(1000, 670)",
  "context.moveTo(0, 693)",
  "context.lineTo(1000, 693)",
  "context.moveTo(0, 716)",
  "context.lineTo(1000, 716)",
  "context.moveTo(0, 739)",
  "context.lineTo(1000, 739)",
  "context.moveTo(0, 762)",
  "context.lineTo(1000, 762)",
  "context.moveTo(0, 785)",
  "context.lineTo(1000, 785)",
  "context.moveTo(0, 808)",
  "context.lineTo(1000, 808)",
  "context.moveTo(0, 831)",
  "context.lineTo(1000, 831)",
  "context.moveTo(0, 854)",
  "context.lineTo(1000, 854)",
  "context.moveTo(0, 877)",
  "context.lineTo(1000, 877)",
  "context.moveTo(0, 900)",
  "context.lineTo(1000, 900)",
  "context.moveTo(0, 923)",
  "context.lineTo(1000, 923)",
  "context.moveTo(0, 946)",
  "context.lineTo(1000, 946)",
  "context.moveTo(0, 969)",
  "context.lineTo(1000, 969)",
  "context.moveTo(0, 992)",
  "context.lineTo(1000, 992)",
  "context.moveTo(0, 1015)",
  "context.lineTo(1000, 1015)",
  "context.stroke()",
  "context.lineWidth=0.12;",
  "context.strokeStyle=\\"#111\\";",
  "context.textBaseline=\\"middle\\";",
  "context.font=\\"400 13px 'Roboto', arial\\";",
  "context.fillStyle=\\"#000\\";",
  "context.textAlign=\\"right\\";",
  "context.save()",
  "context.beginPath()",
  "context.rect(60, 26, 96, 23)",
  "context.clip()",
  "context.fillText(\\"1\\", 153, 39)",
  "context.restore()",
  "context.lineWidth=1.2000000000000002;",
  "context.fillStyle=\\"#f3f7fe\\";",
  "context.fillStyle=\\"#f3f7fe\\";",
  "context.strokeStyle=\\"#3266ca\\";",
  "context.lineWidth=0.6000000000000001;",
  "context.globalCompositeOperation=\\"multiply\\";",
  "context.fillRect(60, 26, 96, 23)",
  "context.strokeRect(60, 26, 96, 23)",
  "context.globalCompositeOperation=\\"source-over\\";",
  "context.strokeStyle=\\"#3266ca\\";",
  "context.lineWidth=1.2000000000000002;",
  "context.strokeRect(60, 26, 96, 23)",
  "context.fillStyle=\\"#F8F9FA\\";",
  "context.font=\\"400 11px 'Roboto', arial\\";",
  "context.textAlign=\\"center\\";",
  "context.textBaseline=\\"middle\\";",
  "context.lineWidth=0.4;",
  "context.strokeStyle=\\"#333\\";",
  "context.fillRect(0, 0, 1000, 26)",
  "context.fillRect(0, 0, 60, 1000)",
  "context.fillStyle=\\"#E8EAED\\";",
  "context.fillStyle=\\"#E8EAED\\";",
  "context.fillRect(60, 0, 96, 26)",
  "context.fillStyle=\\"#E8EAED\\";",
  "context.fillRect(0, 26, 60, 23)",
  "context.beginPath()",
  "context.moveTo(60, 0)",
  "context.lineTo(60, 1000)",
  "context.moveTo(0, 26)",
  "context.lineTo(1000, 26)",
  "context.stroke()",
  "context.beginPath()",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"A\\", 108, 13)",
  "context.moveTo(156, 0)",
  "context.lineTo(156, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"B\\", 204, 13)",
  "context.moveTo(252, 0)",
  "context.lineTo(252, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"C\\", 300, 13)",
  "context.moveTo(348, 0)",
  "context.lineTo(348, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"D\\", 396, 13)",
  "context.moveTo(444, 0)",
  "context.lineTo(444, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"E\\", 492, 13)",
  "context.moveTo(540, 0)",
  "context.lineTo(540, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"F\\", 588, 13)",
  "context.moveTo(636, 0)",
  "context.lineTo(636, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"G\\", 684, 13)",
  "context.moveTo(732, 0)",
  "context.lineTo(732, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"H\\", 780, 13)",
  "context.moveTo(828, 0)",
  "context.lineTo(828, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"I\\", 876, 13)",
  "context.moveTo(924, 0)",
  "context.lineTo(924, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"J\\", 972, 13)",
  "context.moveTo(1020, 0)",
  "context.lineTo(1020, 26)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"1\\", 30, 37.5)",
  "context.moveTo(0, 49)",
  "context.lineTo(60, 49)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"2\\", 30, 60.5)",
  "context.moveTo(0, 72)",
  "context.lineTo(60, 72)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"3\\", 30, 83.5)",
  "context.moveTo(0, 95)",
  "context.lineTo(60, 95)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"4\\", 30, 106.5)",
  "context.moveTo(0, 118)",
  "context.lineTo(60, 118)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"5\\", 30, 129.5)",
  "context.moveTo(0, 141)",
  "context.lineTo(60, 141)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"6\\", 30, 152.5)",
  "context.moveTo(0, 164)",
  "context.lineTo(60, 164)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"7\\", 30, 175.5)",
  "context.moveTo(0, 187)",
  "context.lineTo(60, 187)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"8\\", 30, 198.5)",
  "context.moveTo(0, 210)",
  "context.lineTo(60, 210)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"9\\", 30, 221.5)",
  "context.moveTo(0, 233)",
  "context.lineTo(60, 233)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"10\\", 30, 244.5)",
  "context.moveTo(0, 256)",
  "context.lineTo(60, 256)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"11\\", 30, 267.5)",
  "context.moveTo(0, 279)",
  "context.lineTo(60, 279)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"12\\", 30, 290.5)",
  "context.moveTo(0, 302)",
  "context.lineTo(60, 302)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"13\\", 30, 313.5)",
  "context.moveTo(0, 325)",
  "context.lineTo(60, 325)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"14\\", 30, 336.5)",
  "context.moveTo(0, 348)",
  "context.lineTo(60, 348)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"15\\", 30, 359.5)",
  "context.moveTo(0, 371)",
  "context.lineTo(60, 371)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"16\\", 30, 382.5)",
  "context.moveTo(0, 394)",
  "context.lineTo(60, 394)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"17\\", 30, 405.5)",
  "context.moveTo(0, 417)",
  "context.lineTo(60, 417)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"18\\", 30, 428.5)",
  "context.moveTo(0, 440)",
  "context.lineTo(60, 440)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"19\\", 30, 451.5)",
  "context.moveTo(0, 463)",
  "context.lineTo(60, 463)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"20\\", 30, 474.5)",
  "context.moveTo(0, 486)",
  "context.lineTo(60, 486)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"21\\", 30, 497.5)",
  "context.moveTo(0, 509)",
  "context.lineTo(60, 509)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"22\\", 30, 520.5)",
  "context.moveTo(0, 532)",
  "context.lineTo(60, 532)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"23\\", 30, 543.5)",
  "context.moveTo(0, 555)",
  "context.lineTo(60, 555)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"24\\", 30, 566.5)",
  "context.moveTo(0, 578)",
  "context.lineTo(60, 578)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"25\\", 30, 589.5)",
  "context.moveTo(0, 601)",
  "context.lineTo(60, 601)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"26\\", 30, 612.5)",
  "context.moveTo(0, 624)",
  "context.lineTo(60, 624)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"27\\", 30, 635.5)",
  "context.moveTo(0, 647)",
  "context.lineTo(60, 647)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"28\\", 30, 658.5)",
  "context.moveTo(0, 670)",
  "context.lineTo(60, 670)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"29\\", 30, 681.5)",
  "context.moveTo(0, 693)",
  "context.lineTo(60, 693)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"30\\", 30, 704.5)",
  "context.moveTo(0, 716)",
  "context.lineTo(60, 716)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"31\\", 30, 727.5)",
  "context.moveTo(0, 739)",
  "context.lineTo(60, 739)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"32\\", 30, 750.5)",
  "context.moveTo(0, 762)",
  "context.lineTo(60, 762)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"33\\", 30, 773.5)",
  "context.moveTo(0, 785)",
  "context.lineTo(60, 785)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"34\\", 30, 796.5)",
  "context.moveTo(0, 808)",
  "context.lineTo(60, 808)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"35\\", 30, 819.5)",
  "context.moveTo(0, 831)",
  "context.lineTo(60, 831)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"36\\", 30, 842.5)",
  "context.moveTo(0, 854)",
  "context.lineTo(60, 854)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"37\\", 30, 865.5)",
  "context.moveTo(0, 877)",
  "context.lineTo(60, 877)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"38\\", 30, 888.5)",
  "context.moveTo(0, 900)",
  "context.lineTo(60, 900)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"39\\", 30, 911.5)",
  "context.moveTo(0, 923)",
  "context.lineTo(60, 923)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"40\\", 30, 934.5)",
  "context.moveTo(0, 946)",
  "context.lineTo(60, 946)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"41\\", 30, 957.5)",
  "context.moveTo(0, 969)",
  "context.lineTo(60, 969)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"42\\", 30, 980.5)",
  "context.moveTo(0, 992)",
  "context.lineTo(60, 992)",
  "context.fillStyle=\\"#666666\\";",
  "context.fillText(\\"43\\", 30, 1003.5)",
  "context.moveTo(0, 1015)",
  "context.lineTo(60, 1015)",
  "context.stroke()",
]
`;
